### PR TITLE
[bug-fix] Resource level disable security fix and  test case for disabled security

### PR DESCRIPTION
**Pull request by slahirucd7**
### Purpose

1. 

Allows to disable security in resource level by adding `x-wso2-disable-security` extension. 
- Provides above support in Swagger 2 and OAS 3.

2.

> api level define x-wso2-disable-security: false
 operation level define x-auth-type: None 

Above configuration will disable security in the specified API operation.

### Issues

Fixes https://github.com/wso2/product-microgateway/issues/2485
Fixes https://github.com/wso2/product-microgateway/issues/2337

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes

### Tested environments
java version "11.0.12" 2021-07-20 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.12+8-LTS-237)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.12+8-LTS-237, mixed mode)

go version go1.16.7 darwin/arm64

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

**Review by pubudu538 — APPROVED** [review]

File: adapter/internal/oasparser/model/mgw_swagger.go [hunk 1]
```diff
@@ -838,14 +838,17 @@ func ResolveThrottlingTier(vendorExtensions map[string]interface{}) string {
...
 	z, vExtDisableSecurity := vendorExtensions[xWso2DisableSecurity]
 	if vExtDisableSecurity {
 		// If x-wso2-disable-security is present, then disableSecurity = val
 		if val, ok := z.(bool); ok {
 			disableSecurity = val
+			isXWso2DisableSecurityAvailable = val
 		}
-	} else if vExtAuthType {
+	} 
+	if vExtAuthType && !isXWso2DisableSecurityAvailable {
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/mgw_swagger.go:851]: isnt it always isXWso2DisableSecurityAvailable and disableSecurity is same till this line. So shall we use the disableSecurity here

File: adapter/internal/oasparser/model/open_api.go [hunk 2]
```diff
@@ -125,12 +125,16 @@ func setResourcesOpenAPI(openAPI openapi3.Swagger) ([]Resource, error) {
 	val, found := resolveAPILevelDisableSecurity(openAPI.ExtensionProps)
 	if openAPI.Paths != nil {
 		for path, pathItem := range openAPI.Paths {
+			// Checks for resource level security. (security is disabled in resource level using x-wso2-disable-security extension)
+			isResourceLvlSecurityDisabled, foundInResourceLevel := resolveAPILevelDisableSecurity(pathItem.ExtensionProps)
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/model/open_api.go:129]: shall we rename resolveAPILevelDisableSecurity method as it is not only used to resolve apilevel

File: adapter/internal/oasparser/model/swagger.go [hunk 3]
```diff
@@ -81,9 +81,17 @@ func setResourcesSwagger(swagger2 spec.Swagger) []Resource {
 	// Check if the "x-wso2-disable-security" vendor ext is present at the API level.
 	// If API level vendor ext is present, then the same key:value should be added to
 	// resourve level, if it's not present at resource level using "addResourceLevelDisableSecurity"
-	disableSecurity, found := swagger2.VendorExtensible.Extensions.GetBool(xWso2DisableSecurity)
```
[the remark this answers is not in the document]
Comment (slahirucd7) [reply adapter/internal/oasparser/model/swagger.go:84]: `found` variable is used throughout the loop multiple times. Also the time complexity will be similar with the above suggested approach.

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/extensions/DisableSecurityTestCase.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.HttpStatus;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import org.wso2.choreo.connect.mockbackend.ResponseConstants;
+import org.wso2.choreo.connect.tests.util.*;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class DisableSecurityTestCase {
```
[the remark this answers is not in the document]
Comment (slahirucd7) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/extensions/DisableSecurityTestCase.java:30]: Added test to cover apim scenario.

**Review by pubudu538 — APPROVED** [review]

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[failed] dev-deployment : 20211215.32

**Comment by choreo-cicd** [comment]
[succeeded] dev-deployment : 20211215.43

**Pull request merged**